table/tables/tables.go

```diff
@@ -353,6 +403,41 @@ func (t *Table) getRollbackableMemStore(ctx sessionctx.Context) kv.RetrieverMuta
 	return bs
 }
 
+// locatePartition returns the partition ID of the input record.
+func (t *Table) locatePartition(ctx sessionctx.Context, r []types.Datum) (int64, error) {
+	partitionInfo := t.meta.GetPartitionInfo()
+	if partitionInfo == nil {
```

Comment:
Can we move this judgment outside? We no need to call locatePartition when there are no partitions.

Reply:
Of course we can change, but that change just makes more redundant code.

After the change, the caller write more code:

```
	var pid int64
	partitionInfo := t.meta.GetPartitionInfo()
	if partitionInfo == nil {
		pid = t.ID
	} else {
		var err error
		pid, err = t.locatePartition(ctx, r)
		if err != nil {
			return 0, errors.Trace(err)
		}
	}
```

and the callee still need to call `GetPartitionInfo` because it need it:

```
partitionInfo := t.meta.GetPartitionInfo()
```

the `locatePartition` become more vulnerable because it asserts `partitionInfo` must not be nil.